README.md

```diff
@@ -127,40 +129,100 @@ $sslcert = @("EOT")
-some contexts, such as string interpolation, it might cause surprising results
-(interpolating to `Sensitive [value redacted]`). This is intentional, as it
-makes it more difficult to accidentally mishandle secrets.
+```puppet
+node 'server-123' {
+  $db_password = Deferred(conjur::secret, ['inventory/db-password'])
+
+  file { '/tmp/creds.txt':
+    ensure => file,
+    mode => '0600',
+    content => $db_password,
+  }
+
+  file { '/tmp/creds.ini':
+    ensure => file,
+    mode => '0600',
+    content => Deferred('inline_epp', [
+      'password=<%= $db_password.unwrap %>',
+      { 'db_password' => $db_password }
+    ]),
+  }
+}
+```
 
-To use a secret as a string, you need to explicitly request it using the
-`unwrap` function; the result of the processing should be again wrapped in
-a `Sensitive` value.
+#### `Deferred` functions
 
-In particular, you should not pass unwrapped secrets as resource parameters
-if you can avoid it. Many resource types support `Sensitive` data type and
-handle it correctly. If a resource you're using does not, file a bug.
+This module _requires_ the use of Puppet v6+
+[`Deferred` functions](https://puppet.com/docs/puppet/6.17/deferring_functions.html)
+to ensure that the credential retrieval is fully handled on the agent. Since the
+resolution of variables is done on the agent _after_ the catalog is compiled, anything
+that requires the value of the variable during the compilation step (e.g. template
+compilation)
+[must also be wrapped in a `Deferred` invocation](https://puppet.com/docs/puppet/6.17/template_with_deferred_values.html).
+
+It is also important to note that you should make sure that you invoke the
+`conjur::secret` function using the proper `Deferred` syntax:
+
+```puppet
+# GOOD: Passing the parameters as an array
+Deferred(conjur::secret, ['production/postgres/password'])
+```
+
+```puppet
+# BAD: This will not work!
+Deferred(conjur::secret('production/postgres/password'))
+```
+
+You can read more about Puppet's `Deferred` functions
+[here](https://puppet.com/docs/puppet/6.17/deferring_functions.html).
+
+#### `Sensitive` type
+
+`conjur::secret` returns values wrapped in a `Sensitive` data type. In
+some contexts, such as string interpolation, it might cause surprising results
+(interpolating to `Sensitive [value redacted]`). This is intentional, as it
+makes it more difficult to accidentally mishandle secrets. Additionally, we also
+highly recommend that the whole `Deferred` invocation is also wrapped in
+`Sensitive` type to prevent accidental disclosure of information about the
+function parameters to the logs if more than just the API key is needed to be
+kept private.
+
+To use a `Sensitive` value as a string, you need to explicitly request it using
+the `unwrap` function. In the case of the recommended usage of `conjur::secret`
+and if you are using the value in a template, you need to use
+`$var_name.unwrap` to get the raw string value.
```

Comment:
this could be simplified

I would drop the "In the case of" sentence entirely and just use the example as a guide